client/protocols/collaborate.py

```diff
@@ -0,0 +1,375 @@
+
+        pyrebase = install_and_return('pyrebase')
+        firebase = pyrebase.initialize_app(self.FIREBASE_CONFIG)
+        self.fire_auth = firebase.auth()
+        self.fire_db = firebase.database()
+
+        self.user_email = email
+        self.hostname = platform.node()
+
+        data = {
+            'access_token': access_token,
+            'email': email,
+            'identifier': identifier,
+            'assignment': self.assignment.endpoint,
+            'file_contents': messages.get('file_contents'),
+            'analytics': messages.get('analytics'),
+        }
+
+        # Check for existing sessions first - TBD Future
+        # existing_sessions = self.send_messages(data, endpoint='/collab/list')
+        # response = self.prompt_for_existing_session(existing_sessions.get('sessions'))
+        # if response:
+        #     data['desired_session'] = response
+
+        # Send data to collaborate server
+        response_data = self.send_messages(data, self.LONG_TIMEOUT)
+        if 'error' in response_data or 'session' not in response_data:
+            print("There was an error while starting the session: {} Try again later".format(response_data.get('error')))
+            log.warning("Error: {}".format(response_data.get('error')))
+            return
+
+        self.session_id = response_data['session']
+        self.short_url = response_data['short_url']
+        self.login_user = response_data.get('login_user')
+
+        # Login as the firebase user
+        email, password = response_data.get('login_user'), response_data.get('password')
+
+        try:
+            self.fire_user = self.fire_auth.sign_in_with_email_and_password(email,
+                                                                            password)
+            self.fire_uid = self.fire_user['localId']
+        except (ValueError, KeyError) as e:
+            log.warning("Could not login", exc_info=True)
+            print("Could not login to the collaboration server.")
+            return
+
+        self.stream = (self.get_firebase()
+                           .child('actions').stream(self.stream_listener,
+                                                    self.fire_user['idToken']))
+
+        self.presence = (self.get_firebase()
+                             .child('clients').push({'computer': platform.node(),
+                                                     'uid': self.fire_uid,
+                                                     'owner': self.user_email,
+                                                     'email': self.user_email},
+                                                    self.fire_user['idToken']))
+
+        # Parse response_url
+        if response_data:
+            open_url = response_data['url']
+            if 'access_token' not in open_url:
+                open_url = open_url + "?access_token={}".format(access_token)
+            webbrowser.open_new(open_url)
```

Comment:
This might not work on some systems, in which case it might be worth printing the URL (as you do below) instead of failing (in case we implement the copy-paste log in flow for those systems)